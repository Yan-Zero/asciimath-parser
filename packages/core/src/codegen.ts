import type { ChildNode, MatrixNode, RootNode } from './parser'
import { NodeTypes } from './parser'

function getArrayBoundary(node: MatrixNode) {
  const div = node.dividerIndices
  let beginArray = '\\begin{array}'
  const ch = node.alignment

  if (div.length) {
    const divMax = div[div.length - 1]
    // transform divider indices to col numbers between two divider
    // [2, 4, 5] -> [2, 2, 1]
    for (let i = div.length - 1; i >= 1; i--)
      div[i] -= div[i - 1]

    // [2, 2, 1] -> {cc|cc|c|
    beginArray += '{'
    for (let i = 0; i < div.length; i++)
      beginArray += `${ch.repeat(div[i])}|`

    // MathJax would complain if the array env arg
    // is not consistent with the number of elements of the matrix row.
    // For example, the matrix is like
    // [a, b | c;
    //  d, e | f]
    // if the array env arg is `\begin{array}{cc|}`, the bar won't render correctly.
    // Change it to `\begin{array}{cc|c}` should fix this problem.
    const maxCol = Math.max(...node.params.map(i => i.length))
    // {cc|cc|c| -> {cc|cc|c|...c}
    beginArray += `${ch.repeat(maxCol - divMax)}}`
  }
  else {
    const maxCol = Math.max(...node.params.map(c => c.length))
    beginArray += `{${ch.repeat(maxCol)}}`
  }
  return [beginArray, '\\end{array}']
}

function codegen(node: ChildNode | RootNode): string {
  switch (node.type) {
    case NodeTypes.Const: {
      return node.tex
    }
    case NodeTypes.Root: {
      let res = node.body.map(codegen).join(' ')
      if (node.body.find(n => n.type === NodeTypes.Const && (n.value === '&' || n.tex === '\\\\')))
        res = `\\begin{aligned}${res}\\end{aligned}`
      return res
    }
    case NodeTypes.Flat: {
      return node.body.map(codegen).join(' ')
    }
    case NodeTypes.Matrix: {
      const [arrayBegin, arrayEnd] = getArrayBoundary(node)
      return [
        node.lparen,
        arrayBegin,
        node.params.map(i => i.map(codegen).join(' & ')).join('\\\\'),
        '\\\\',
        arrayEnd,
        node.rparen,
      ].join(' ')
    }
    case NodeTypes.ParamOne: {
      return node.tex.replace('$1', codegen(node.params))
    }
    case NodeTypes.ParamTwo: {
      return node.tex.replace('$1', codegen(node.params[0])).replace('$2', codegen(node.params[1]))
    }
  }
}

export {
  codegen,
}
