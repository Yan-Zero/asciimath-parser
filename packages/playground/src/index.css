@import url('https://fonts.googleapis.com/css2?family=JetBrains+Mono&display=swap');

:root {
  font-family: Inter, Avenir, Helvetica, Arial, sans-serif;
  font-size: 16px;
  line-height: 24px;
  font-weight: 400;

  color-scheme: light dark;
  color: rgba(255, 255, 255, 0.87);
  background-color: #242424;

  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-size-adjust: 100%;
}

a {
  font-weight: 500;
  color: #646cff;
  text-decoration: inherit;
}
a:hover {
  color: #535bf2;
}

body {
  margin: 0;
  min-width: 320px;
  min-height: 100vh;
}

h1 {
  font-size: 3.2em;
  line-height: 1.1;
  margin-block-end: 0.25rem;
}

h2 {
  margin-top: 3rem;
}

h3 {
  margin-top: 2rem;
}

button {
  border-radius: 8px;
  border: 1px solid transparent;
  padding: 0.6em 1.2em;
  font-size: 1em;
  font-weight: 500;
  font-family: inherit;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.25s;
  display: flex;
  align-items: center;
  justify-content: center;
}
button:hover {
  border-color: #646cff;
}
button:focus,
button:focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

@media (prefers-color-scheme: light) {
  :root {
    color: #213547;
    background-color: #ffffff;
  }
  a:hover {
    color: #747bff;
  }
  button {
    background-color: #f9f9f9;
  }
}

td {
  padding: 0.5rem;
  white-space: nowrap;
}

table {
  width: 100%;
  border-top: 2px solid;
  border-bottom: 2px solid;
}

.tc {
  text-align: center;
}

.tex {
  background-color: #7f7f7f33;
}

main {
  max-width: 800px;
  margin: 0 auto;
  padding: 1rem;
}

.space-x-1 > :not([hidden]) ~ :not([hidden]) {
  --tw-space-x-reverse: 0;
  margin-right: calc(0.25rem * var(--tw-space-x-reverse));
  margin-left: calc(0.25rem * calc(1 - var(--tw-space-x-reverse)));
}

.flex {
  display: flex;
}
.w-full {
  width: 100%;
}
.mb-4 {
  margin-bottom: 1rem;
}

code, pre {
  font-family: 'JetBrains Mono', 'Courier New', Courier, monospace;
  font-size: 0.8rem;
}

code {
  background-color: #7f7f7f11;
  padding: 0.2rem;
  border-radius: 0.2rem;
}

pre {
  background-color: #7f7f7f11;
  padding: 0.75em;
  border-radius: 0.2rem;
  overflow: auto;
}