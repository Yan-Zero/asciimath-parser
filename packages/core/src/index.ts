import { codegen } from './codegen'
import { parser } from './parser'
import type { Trie } from './trie'
import { createTrie } from './trie'

type ReplaceLaw = [RegExp | string, string | ((substring: string, ...args: any[]) => string)]

interface AsciiMathConfig {
  /**
   * @default true
   * enable displayMode in KaTeX
   */
  display?: boolean
  /**
   * Translate custom keywords into LaTeX expressions
   *
   * For example:
   * [
   *   ['dx', '\text{d}x'],
   *   ['dy', '\text{d}y']
   * ]
   */
  extConst?: Array<[string, string]>
  /**
   * Replace target expressions before tokenizing
   *
   * For example:
   * [
   *   [/&#(x?[0-9a-fA-F]+);/g, (match, $1) =>
   *     String.fromCodePoint($1[0] === 'x' ? '0' + $1 : $1)
   *   ],
   *   ...
   * ]
   */
  replaceBeforeTokenizing?: ReplaceLaw[]
}

type RestrictedAmConfig = Required<AsciiMathConfig>

function resolveConfig(config?: AsciiMathConfig): RestrictedAmConfig {
  const defaultConfig: RestrictedAmConfig = {
    display: true,
    extConst: [
      ['dx', '{\\text{d}x}'],
      ['dy', '{\\text{d}y}'],
      ['dz', '{\\text{d}z}'],
      ['dt', '{\\text{d}t}'],
      ['#', '\\displaystyle'],
    ],
    replaceBeforeTokenizing: [
      [/&#(x?[0-9a-fA-F]+);/g, (_match, $1) =>
        String.fromCodePoint($1[0] === 'x' ? `0${$1}` : $1),
      ],
    ],
  }
  if (typeof config?.display !== 'undefined')
    defaultConfig.display = config?.display
  if (config?.extConst?.length)
    defaultConfig.extConst.push(...config.extConst)
  if (config?.replaceBeforeTokenizing?.length)
    defaultConfig.replaceBeforeTokenizing.push(...config.replaceBeforeTokenizing)

  return defaultConfig
}

class AsciiMath {
  private trie: Trie
  private display: boolean
  private replaceLaws: ReplaceLaw[]
  constructor(config?: AsciiMathConfig) {
    const { display, extConst, replaceBeforeTokenizing: replaceBeforeParsing } = resolveConfig(config)
    this.trie = createTrie({ extConst })
    this.display = display
    this.replaceLaws = replaceBeforeParsing
  }

  toTex(code: string): string {
    try {
      code = this.replaceLaws.reduce((prev, curLaw) => {
      // in fact the judgement is no use...
        if (typeof curLaw[1] === 'function')
          return prev.replace(curLaw[0], curLaw[1])
        else
          return prev.replace(curLaw[0], curLaw[1])
      }, code)
      let res = codegen(parser(this.trie.tryParsingAll(code)))
      if (this.display)
        res = `\\displaystyle{ ${res} }`
      return res
    }
    catch (e) {
      return `\\text{${String(e)}}`
    }
  }
}

export {
  AsciiMath,
  AsciiMathConfig,
}
