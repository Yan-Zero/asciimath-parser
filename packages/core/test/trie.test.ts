import { describe, expect, it } from 'vitest'
import { Trie, createTrie } from '../src/trie'

const MULTILINE_AM = `sum_(n=1)^(+oo)&=(pi^2/6)

f(x)&=x^2`

describe('trie', () => {
  it('should build trie', () => {
    const keys = Array.from({ length: 26 }, (_, i) => String.fromCharCode(i + 97))
    keys.push(' ')
    const trie = new Trie(keys)
    trie.insert('banana')
    trie.insert('ban')
    trie.insert('abc')
    trie.insert('abda')

    expect(trie).toMatchSnapshot()
    expect(trie.search('ban')).toBe(true)
    expect(trie.search('aaaa')).toBe(false)
    // directly search keyword
    expect(trie.tryParsing('banana')).toMatchInlineSnapshot(`
      {
        "current": 6,
        "isKeyWord": true,
        "value": "banana",
      }
    `)
    // match the longest keyword and return the keyword
    expect(trie.tryParsing('bananan')).toMatchInlineSnapshot(`
      {
        "current": 6,
        "isKeyWord": true,
        "value": "banana",
      }
    `)
    // it does not match any keyword
    expect(trie.tryParsing('bab')).toMatchInlineSnapshot(`
      {
        "current": 2,
        "isKeyWord": false,
        "tex": "ba",
        "type": "StringLiteral",
        "value": "ba",
      }
    `)
    // skip when meeting a space
    expect(trie.tryParsing('ab cfdgerafw')).toMatchInlineSnapshot(`
      {
        "current": 2,
        "isKeyWord": false,
        "tex": "ab",
        "type": "StringLiteral",
        "value": "ab",
      }
    `)
    // maybe it is somewhat strange...
    expect(trie.tryParsing('bana')).toMatchInlineSnapshot(`
      {
        "current": 4,
        "isKeyWord": false,
        "tex": "bana",
        "type": "StringLiteral",
        "value": "bana",
      }
    `)
  })
})

describe('asciimath matrix', () => {
  const trie = createTrie()
  it('should build the symbol trie', () => {
    expect(trie.tryParsingAll(MULTILINE_AM)).toMatchSnapshot()
    expect(trie.tryParsingAll('x |-> "e"^(2pi "i" x)')).toMatchSnapshot()
    expect(trie.tryParsingAll('"e"^("i" pi)')).toMatchSnapshot()
  })

  it('should tokenize a matrix', () => {
    expect(trie.tryParsingAll('[a,b,c;d,e;f}')).toMatchSnapshot()
  })

  it('should tokenize a recursive matrix', () => {
    expect(trie.tryParsingAll('[[a,b],c;d,e:}')).toMatchSnapshot()
  })

  it('should tokenize a divided matrix', () => {
    const m = `[
      a, b, |, c;
      d, e, |, f;
      g, | h, i
    ]`
    expect(trie.tryParsingAll(m)).toMatchSnapshot()
  })

  it('should tokenize cases', () => {
    expect(trie.tryParsingAll('f(x)={x^2,if x>0;x, otherwise:}')).toMatchSnapshot()
  })
})
