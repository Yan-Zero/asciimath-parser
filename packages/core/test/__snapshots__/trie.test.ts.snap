// Vitest Snapshot v1

exports[`asciimath matrix > should build the symbol trie 1`] = `
[
  {
    "current": 3,
    "isKeyWord": true,
    "tex": "\\\\sum",
    "type": "Const",
    "value": "sum",
  },
  {
    "current": 4,
    "isKeyWord": true,
    "tex": "_{ $1 }",
    "type": "OperatorSup",
    "value": "_",
  },
  {
    "current": 5,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 6,
    "isKeyWord": false,
    "tex": "n",
    "type": "StringLiteral",
    "value": "n",
  },
  {
    "current": 7,
    "isKeyWord": false,
    "tex": "=",
    "type": "StringLiteral",
    "value": "=",
  },
  {
    "current": 8,
    "isKeyWord": false,
    "tex": "1",
    "type": "NumberLiteral",
    "value": "1",
  },
  {
    "current": 9,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
  {
    "current": 10,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorSup",
    "value": "^",
  },
  {
    "current": 11,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 12,
    "isKeyWord": true,
    "tex": "{+$1 }",
    "type": "OperatorMinus",
    "value": "+",
  },
  {
    "current": 14,
    "isKeyWord": true,
    "tex": "\\\\infty",
    "type": "Const",
    "value": "oo",
  },
  {
    "current": 15,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
  {
    "current": 16,
    "isKeyWord": false,
    "tex": "1",
    "type": "NumberLiteral",
    "value": "1",
  },
  {
    "current": 17,
    "isKeyWord": true,
    "tex": "\\\\frac{ $1 }{ $2 }",
    "type": "OperatorAOB",
    "value": "/",
  },
  {
    "current": 18,
    "isKeyWord": false,
    "tex": "n",
    "type": "StringLiteral",
    "value": "n",
  },
  {
    "current": 19,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorSup",
    "value": "^",
  },
  {
    "current": 20,
    "isKeyWord": false,
    "tex": "2",
    "type": "NumberLiteral",
    "value": "2",
  },
  {
    "current": 21,
    "isKeyWord": true,
    "tex": "&",
    "type": "Align",
    "value": "&",
  },
  {
    "current": 22,
    "isKeyWord": false,
    "tex": "=",
    "type": "StringLiteral",
    "value": "=",
  },
  {
    "current": 24,
    "isKeyWord": true,
    "tex": "\\\\pi",
    "type": "Const",
    "value": "pi",
  },
  {
    "current": 25,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorSup",
    "value": "^",
  },
  {
    "current": 26,
    "isKeyWord": false,
    "tex": "2",
    "type": "NumberLiteral",
    "value": "2",
  },
  {
    "current": 27,
    "isKeyWord": true,
    "tex": "\\\\frac{ $1 }{ $2 }",
    "type": "OperatorAOB",
    "value": "/",
  },
  {
    "current": 28,
    "isKeyWord": false,
    "tex": "6",
    "type": "NumberLiteral",
    "value": "6",
  },
  {
    "current": 30,
    "isKeyWord": true,
    "tex": "\\\\\\\\",
    "type": "Align",
    "value": "

",
  },
  {
    "current": 31,
    "isKeyWord": false,
    "tex": "f",
    "type": "StringLiteral",
    "value": "f",
  },
  {
    "current": 32,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 33,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 34,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
  {
    "current": 35,
    "isKeyWord": true,
    "tex": "&",
    "type": "Align",
    "value": "&",
  },
  {
    "current": 36,
    "isKeyWord": false,
    "tex": "=",
    "type": "StringLiteral",
    "value": "=",
  },
  {
    "current": 37,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 38,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorSup",
    "value": "^",
  },
  {
    "current": 39,
    "isKeyWord": false,
    "tex": "2",
    "type": "NumberLiteral",
    "value": "2",
  },
]
`;

exports[`asciimath matrix > should build the symbol trie 2`] = `
[
  {
    "current": 1,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 5,
    "isKeyWord": true,
    "tex": "\\\\mapsto",
    "type": "Const",
    "value": "|->",
  },
  {
    "current": 9,
    "isKeyWord": false,
    "tex": "e",
    "type": "Text",
    "value": "e",
  },
  {
    "current": 10,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorSup",
    "value": "^",
  },
  {
    "current": 11,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 12,
    "isKeyWord": false,
    "tex": "2",
    "type": "NumberLiteral",
    "value": "2",
  },
  {
    "current": 14,
    "isKeyWord": true,
    "tex": "\\\\pi",
    "type": "Const",
    "value": "pi",
  },
  {
    "current": 18,
    "isKeyWord": false,
    "tex": "i",
    "type": "Text",
    "value": "i",
  },
  {
    "current": 20,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 21,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
]
`;

exports[`asciimath matrix > should build the symbol trie 3`] = `
[
  {
    "current": 3,
    "isKeyWord": false,
    "tex": "e",
    "type": "Text",
    "value": "e",
  },
  {
    "current": 4,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorSup",
    "value": "^",
  },
  {
    "current": 5,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 8,
    "isKeyWord": false,
    "tex": "i",
    "type": "Text",
    "value": "i",
  },
  {
    "current": 11,
    "isKeyWord": true,
    "tex": "\\\\pi",
    "type": "Const",
    "value": "pi",
  },
  {
    "current": 12,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
]
`;

exports[`asciimath matrix > should tokenize a divided matrix 1`] = `
[
  {
    "current": 1,
    "isKeyWord": true,
    "tex": "[",
    "type": "LParen",
    "value": "[",
  },
  {
    "current": 9,
    "isKeyWord": false,
    "tex": "a",
    "type": "StringLiteral",
    "value": "a",
  },
  {
    "current": 10,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 12,
    "isKeyWord": false,
    "tex": "b",
    "type": "StringLiteral",
    "value": "b",
  },
  {
    "current": 13,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 15,
    "isKeyWord": true,
    "tex": "|",
    "type": "Paren",
    "value": "|",
  },
  {
    "current": 16,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 18,
    "isKeyWord": false,
    "tex": "c",
    "type": "StringLiteral",
    "value": "c",
  },
  {
    "current": 19,
    "isKeyWord": true,
    "tex": ";",
    "type": "Split",
    "value": ";",
  },
  {
    "current": 27,
    "isKeyWord": false,
    "tex": "d",
    "type": "StringLiteral",
    "value": "d",
  },
  {
    "current": 28,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 30,
    "isKeyWord": false,
    "tex": "e",
    "type": "StringLiteral",
    "value": "e",
  },
  {
    "current": 31,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 33,
    "isKeyWord": true,
    "tex": "|",
    "type": "Paren",
    "value": "|",
  },
  {
    "current": 34,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 36,
    "isKeyWord": false,
    "tex": "f",
    "type": "StringLiteral",
    "value": "f",
  },
  {
    "current": 37,
    "isKeyWord": true,
    "tex": ";",
    "type": "Split",
    "value": ";",
  },
  {
    "current": 45,
    "isKeyWord": false,
    "tex": "g",
    "type": "StringLiteral",
    "value": "g",
  },
  {
    "current": 46,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 48,
    "isKeyWord": true,
    "tex": "|",
    "type": "Paren",
    "value": "|",
  },
  {
    "current": 50,
    "isKeyWord": false,
    "tex": "h",
    "type": "StringLiteral",
    "value": "h",
  },
  {
    "current": 51,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 53,
    "isKeyWord": false,
    "tex": "i",
    "type": "StringLiteral",
    "value": "i",
  },
  {
    "current": 59,
    "isKeyWord": true,
    "tex": "]",
    "type": "RParen",
    "value": "]",
  },
]
`;

exports[`asciimath matrix > should tokenize a matrix 1`] = `
[
  {
    "current": 1,
    "isKeyWord": true,
    "tex": "[",
    "type": "LParen",
    "value": "[",
  },
  {
    "current": 2,
    "isKeyWord": false,
    "tex": "a",
    "type": "StringLiteral",
    "value": "a",
  },
  {
    "current": 3,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 4,
    "isKeyWord": false,
    "tex": "b",
    "type": "StringLiteral",
    "value": "b",
  },
  {
    "current": 5,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 6,
    "isKeyWord": false,
    "tex": "c",
    "type": "StringLiteral",
    "value": "c",
  },
  {
    "current": 7,
    "isKeyWord": true,
    "tex": ";",
    "type": "Split",
    "value": ";",
  },
  {
    "current": 8,
    "isKeyWord": false,
    "tex": "d",
    "type": "StringLiteral",
    "value": "d",
  },
  {
    "current": 9,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 10,
    "isKeyWord": false,
    "tex": "e",
    "type": "StringLiteral",
    "value": "e",
  },
  {
    "current": 11,
    "isKeyWord": true,
    "tex": ";",
    "type": "Split",
    "value": ";",
  },
  {
    "current": 12,
    "isKeyWord": false,
    "tex": "f",
    "type": "StringLiteral",
    "value": "f",
  },
  {
    "current": 13,
    "isKeyWord": true,
    "tex": "\\\\rbrace",
    "type": "RParen",
    "value": "}",
  },
]
`;

exports[`asciimath matrix > should tokenize a recursive matrix 1`] = `
[
  {
    "current": 1,
    "isKeyWord": true,
    "tex": "[",
    "type": "LParen",
    "value": "[",
  },
  {
    "current": 2,
    "isKeyWord": true,
    "tex": "[",
    "type": "LParen",
    "value": "[",
  },
  {
    "current": 3,
    "isKeyWord": false,
    "tex": "a",
    "type": "StringLiteral",
    "value": "a",
  },
  {
    "current": 4,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 5,
    "isKeyWord": false,
    "tex": "b",
    "type": "StringLiteral",
    "value": "b",
  },
  {
    "current": 6,
    "isKeyWord": true,
    "tex": "]",
    "type": "RParen",
    "value": "]",
  },
  {
    "current": 7,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 8,
    "isKeyWord": false,
    "tex": "c",
    "type": "StringLiteral",
    "value": "c",
  },
  {
    "current": 9,
    "isKeyWord": true,
    "tex": ";",
    "type": "Split",
    "value": ";",
  },
  {
    "current": 10,
    "isKeyWord": false,
    "tex": "d",
    "type": "StringLiteral",
    "value": "d",
  },
  {
    "current": 11,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 12,
    "isKeyWord": false,
    "tex": "e",
    "type": "StringLiteral",
    "value": "e",
  },
  {
    "current": 14,
    "isKeyWord": true,
    "tex": ".",
    "type": "RParen",
    "value": ":}",
  },
]
`;

exports[`asciimath matrix > should tokenize cases 1`] = `
[
  {
    "current": 1,
    "isKeyWord": false,
    "tex": "f",
    "type": "StringLiteral",
    "value": "f",
  },
  {
    "current": 2,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 3,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 4,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
  {
    "current": 5,
    "isKeyWord": false,
    "tex": "=",
    "type": "StringLiteral",
    "value": "=",
  },
  {
    "current": 6,
    "isKeyWord": true,
    "tex": "\\\\lbrace",
    "type": "LParen",
    "value": "{",
  },
  {
    "current": 7,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 8,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorSup",
    "value": "^",
  },
  {
    "current": 9,
    "isKeyWord": false,
    "tex": "2",
    "type": "NumberLiteral",
    "value": "2",
  },
  {
    "current": 10,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 12,
    "isKeyWord": true,
    "tex": "if\\\\quad",
    "type": "Text",
    "value": "if",
  },
  {
    "current": 14,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 15,
    "isKeyWord": false,
    "tex": ">",
    "type": "StringLiteral",
    "value": ">",
  },
  {
    "current": 16,
    "isKeyWord": false,
    "tex": "0",
    "type": "NumberLiteral",
    "value": "0",
  },
  {
    "current": 17,
    "isKeyWord": true,
    "tex": ";",
    "type": "Split",
    "value": ";",
  },
  {
    "current": 18,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 19,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 29,
    "isKeyWord": true,
    "tex": "otherwise\\\\quad",
    "type": "Text",
    "value": "otherwise",
  },
  {
    "current": 31,
    "isKeyWord": true,
    "tex": ".",
    "type": "RParen",
    "value": ":}",
  },
]
`;

exports[`tokenize color > should tokenize color 1`] = `
[
  {
    "current": 5,
    "isKeyWord": true,
    "tex": "{ \\\\color{$1} $2 }",
    "type": "OperatorOAB",
    "value": "color",
  },
  {
    "current": 11,
    "isKeyWord": false,
    "tex": "pink",
    "type": "Const",
    "value": "pink",
  },
  {
    "current": 12,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 15,
    "isKeyWord": false,
    "tex": "123",
    "type": "NumberLiteral",
    "value": "123",
  },
  {
    "current": 16,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
]
`;

exports[`tokenize color > should tokenize rgb color 1`] = `
[
  {
    "current": 5,
    "isKeyWord": true,
    "tex": "{ \\\\color{$1} $2 }",
    "type": "OperatorOAB",
    "value": "color",
  },
  {
    "current": 14,
    "isKeyWord": false,
    "tex": "#114514",
    "type": "Const",
    "value": "#114514",
  },
  {
    "current": 15,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 18,
    "isKeyWord": false,
    "tex": "123",
    "type": "NumberLiteral",
    "value": "123",
  },
  {
    "current": 19,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
]
`;

exports[`trie success > should build trie 1`] = `
Trie {
  "_char_to_index": Map {
    "a" => 0,
    "b" => 1,
    "c" => 2,
    "d" => 3,
    "e" => 4,
    "f" => 5,
    "g" => 6,
    "h" => 7,
    "i" => 8,
    "j" => 9,
    "k" => 10,
    "l" => 11,
    "m" => 12,
    "n" => 13,
    "o" => 14,
    "p" => 15,
    "q" => 16,
    "r" => 17,
    "s" => 18,
    "t" => 19,
    "u" => 20,
    "v" => 21,
    "w" => 22,
    "x" => 23,
    "y" => 24,
    "z" => 25,
    " " => 26,
  },
  "_n": 27,
  "_root": TrieNode {
    "_end": false,
    "_nextNode": [
      TrieNode {
        "_end": false,
        "_nextNode": [
          null,
          TrieNode {
            "_end": false,
            "_nextNode": [
              null,
              null,
              TrieNode {
                "_end": true,
                "_nextNode": [
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                ],
              },
              TrieNode {
                "_end": false,
                "_nextNode": [
                  TrieNode {
                    "_end": true,
                    "_nextNode": [
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                    ],
                  },
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                ],
              },
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
            ],
          },
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
      },
      TrieNode {
        "_end": false,
        "_nextNode": [
          TrieNode {
            "_end": false,
            "_nextNode": [
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              TrieNode {
                "_end": true,
                "_nextNode": [
                  TrieNode {
                    "_end": false,
                    "_nextNode": [
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      TrieNode {
                        "_end": false,
                        "_nextNode": [
                          TrieNode {
                            "_end": true,
                            "_nextNode": [
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                              null,
                            ],
                          },
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                          null,
                        ],
                      },
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                    ],
                  },
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                ],
              },
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
            ],
          },
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
      },
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
    ],
  },
}
`;
